src/Agent.Worker/ContainerOperationProvider.cs

```diff
@@ -191,8 +191,32 @@ private async Task PullContainerAsync(IExecutionContext executionContext, Contai
                 ArgUtil.NotNullOrEmpty(username, nameof(username));
                 ArgUtil.NotNullOrEmpty(password, nameof(password));
 
-                int loginExitCode = await _dockerManger.DockerLogin(executionContext, registryServer, username, password);
-                if (loginExitCode != 0)
+
+                // login with retry up to 3 times
+                int maxRetries = 3;
```

Comment:
[suggestion/non-blocking] It would be cleaner to encapsulate the retry logic into _dockerManger implementation.

Reply:
Understandable, just following myself the previous author -> Ting's code, would be ideal a global retry mechanism though!